staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go

```diff
@@ -0,0 +1,110 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package external_metrics
+
+import (
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	serializer "k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/metrics/pkg/apis/external_metrics/v1beta1"
+)
+
+type externalMetricsClient struct {
+	client rest.Interface
+}
+
+func New(client rest.Interface) ExternalMetricsClient {
+	return &externalMetricsClient{
+		client: client,
+	}
+}
+
+func NewForConfig(c *rest.Config) (ExternalMetricsClient, error) {
+	configShallowCopy := *c
+	if configShallowCopy.RateLimiter == nil && configShallowCopy.QPS > 0 {
+		configShallowCopy.RateLimiter = flowcontrol.NewTokenBucketRateLimiter(configShallowCopy.QPS, configShallowCopy.Burst)
+	}
+	configShallowCopy.APIPath = "/apis"
+	if configShallowCopy.UserAgent == "" {
+		configShallowCopy.UserAgent = rest.DefaultKubernetesUserAgent()
+	}
+	configShallowCopy.GroupVersion = &v1beta1.SchemeGroupVersion
+	configShallowCopy.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: scheme.Codecs}
+
+	client, err := rest.RESTClientFor(&configShallowCopy)
+	if err != nil {
+		return nil, err
+	}
+
+	return New(client), nil
+}
+
+func NewForConfigOrDie(c *rest.Config) ExternalMetricsClient {
+	client, err := NewForConfig(c)
+	if err != nil {
+		panic(err)
+	}
+	return client
+}
+
+func (c *externalMetricsClient) NamespacedMetrics(namespace string) MetricsInterface {
+	return &namespacedMetrics{
+		client:    c,
+		namespace: namespace,
+	}
+}
+
+type namespacedMetrics struct {
+	client    *externalMetricsClient
+	namespace string
+}
+
+func (m *namespacedMetrics) List(metricName string, metricSelector labels.Selector) (*v1beta1.ExternalMetricValueList, error) {
+	// Slashes can't be used in REST paths, which is a shame
+	// as some metrics system use them for namespacing.
+	// As per https://github.com/kubernetes/community/pull/1802
+	// we escape the slashes to workaround this issue.
+	escapedName := strings.Replace(metricName, "/", "\\|", -1)
+	res := &v1beta1.ExternalMetricValueList{}
+	err := m.client.client.Get().
+		Namespace(m.namespace).
+		Resource(escapedName).
+		VersionedParams(&metav1.ListOptions{
+			LabelSelector: metricSelector.String(),
+		}, metav1.ParameterCodec).
+		Do().
+		Into(res)
+
+	if err != nil {
+		return nil, err
+	}
+
+	newItems := []v1beta1.ExternalMetricValue{}
+	for _, emv := range res.Items {
+		newEMV := emv.DeepCopy()
+		newEMV.MetricName = strings.Replace(emv.MetricName, "\\|", "/", -1)
```

Comment:
requiring client-side fixup for the field that is supposed to uniquely identify the object looks very strange

Reply:
agreed.  I very much dislike this escape.

Reply:
removed